sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -73,28 +80,25 @@ case class BroadcastExchangeExec(
       // with the correct execution.
       SQLExecution.withExecutionId(sparkContext, executionId) {
         try {
-          val beforeCollect = System.nanoTime()
-          // Note that we use .executeCollect() because we don't want to convert data to Scala types
-          val input: Array[InternalRow] = child.executeCollect()
-          if (input.length >= 512000000) {
-            throw new SparkException(
-              s"Cannot broadcast the table with more than 512 millions rows: ${input.length} rows")
-          }
           val beforeBuild = System.nanoTime()
-          longMetric("collectTime") += (beforeBuild - beforeCollect) / 1000000
-          val dataSize = input.map(_.asInstanceOf[UnsafeRow].getSizeInBytes.toLong).sum
-          longMetric("dataSize") += dataSize
-          if (dataSize >= (8L << 30)) {
+          // Call persist on the RDD because we want to broadcast the RDD blocks on executors.
+          childRDD = child.execute().mapPartitionsInternal { rowIterator =>
+            rowIterator.map(_.copy())
```

Comment:
I know the PR has been closed, but I was interested in understanding the code.
Why is copy of `child` RDD made before persisting ?

Reply:
In SparkSQL, the underlying mechanism reuses single object (the row here) when iterating. We must manually copy the row in `map` before persisting. Otherwise you will get the results with all same rows.